### DISCUSSION TITLE
lnd should scan across account/index BIP32 derivations

### DISCUSSION THREAD

**Discussion by alecov** [lightningnetwork/lnd#7789] [Troubleshooting]
From `lncli wallet accounts list` I gather lnd is scanning the following derivations:

```
m/49'/0'/0' (segwit p2sh)
m/84'/0'/0' (segwit native)
m/86'/0'/0' (taproot)
m/1017'/0'/X' (???)
```

Why doesn't it scan the standard account/index at each level, eg. `m/84'/0'/0'/{0,1}/X`?

lnd misses many wallet transactions (account = 0) and wallet changes (at account = 1) since it only looks at the root address (which has usually nothing).

**Comment by guggero** [comment]
`lnd` itself does not use any accounts other than account number `0` internally, at least for any standard key scope (1017 is the `lnd` internal scope for channel/backup/watchtower related keys, there the account number is called the key family, see https://github.com/lightningnetwork/lnd/blob/master/keychain/derivation.go#L67.

The "account" level is the third item in the path. So when you refer to `m/84'/0'/0'/{0,1}/X`, the `{0,1}` is the change (which as per BIP044 is always only 0 or 1) and `X` is the key index. `lnd` does scan more than just index 0, the value `recovery_window` in [`InitWallet`](https://lightning.engineering/api-docs/api/lnd/wallet-unlocker/init-wallet) dictates how many keys.

With all that said, I'm not sure I fully understand your question. `lnd` should not "miss" any keys, if they were created with `lnd` in the first place. It is not recommended to use a seed created by `lnd` with any other wallet.

**Reply by alecov** [reply]
> With all that said, I'm not sure I fully understand your question.
> It is not recommended to use a seed created by lnd with any other wallet.

Why not? Is there a technical reason for this?

I ask specifically why `lnd` does not scan for `m/84'/0'/0'/0/X` and `m/84'/0'/0'/1/X` (for changes) as most or every other wallet software does. This kind of interop is important. Missing that makes it necessary to on-chain transfer between standard BIP32 and lnd-specific wallets.

**Reply by guggero** [reply]
`lnd` **does** scan for addresses on the `m/84'/0'/0'/0/X` path. But I guess the question is: **When** does `lnd` scan for addresses. During normal operation, `lnd` will only keep track of addresses up to index `X` for which it has derived addresses by itself (meaning you called `lncli newaddress` X number of times). So if you derive `X+1` in another wallet and send funds to it without also calling `lncli newaddress` first, `lnd` will not "see" the funds on that address.
But: If you do a wallet restore from seed, then `lnd` will scan for `recovery_window` (so 2500 by default) addresses, so up to `m/84'/0'/0'/0/2500`.

**Reply by alecov** [reply]
>During normal operation, lnd will only keep track of addresses up to index X for which it has derived addresses by itself (meaning you called lncli newaddress X number of times).

BIP-44 suggests an address gap of 20, so I believe it would make sense for lnd to scan 20 addresses from the latest index X known to lnd during boot (not from initial restore, but after shutdown & restart) and set X accordingly.

>But: If you do a wallet restore from seed, then lnd will scan for recovery_window (so 2500 by default) addresses, so up to m/84'/0'/0'/0/2500.

Thanks for clarifying. I was aware of `recovery_window` but still, I've restored from a root xprv and `lnd` shows a zero wallet balance. Does the above scan apply for xprv recovery also?

Also, is the seed recovery option BIP-39 compliant or a different scheme? Are 12-word phrases supported?

**Reply by guggero** [reply]
Unfortunately the [`btcwallet`](https://github.com/btcsuite/btcwallet) library we use for the on-chain handling is a bit dated and doesn't have that many options for address derivation and watching them on chain.

`lnd` uses the [`aezeed`](https://github.com/lightningnetwork/lnd/tree/master/aezeed) format which uses BIP39 words to encode the seed data as a mnemonic phrase. But the content is not compatible with other seeds.
